hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/key/OmKeysDeleteRequestWithFSO.java

```diff
@@ -0,0 +1,266 @@
+
+/**
+ * Handles DeleteKeys request for recursive bucket deletion.
+ */
+public class OmKeysDeleteRequestWithFSO extends OMKeysDeleteRequest {
+
+  private static final Logger LOG =
+      LoggerFactory.getLogger(OmKeysDeleteRequestWithFSO.class);
+
+  public OmKeysDeleteRequestWithFSO(
+      OzoneManagerProtocolProtos.OMRequest omRequest,
+      BucketLayout bucketLayout) {
+    super(omRequest, bucketLayout);
+  }
+
+
+  @Override
+  @SuppressWarnings("methodlength")
+  public OMClientResponse validateAndUpdateCache(OzoneManager ozoneManager,
+      long trxnLogIndex, OzoneManagerDoubleBufferHelper omDoubleBufferHelper) {
+    OzoneManagerProtocolProtos.DeleteKeysRequest deleteKeyRequest =
+        getOmRequest().getDeleteKeysRequest();
+
+    OzoneManagerProtocolProtos.DeleteKeyArgs deleteKeyArgs =
+        deleteKeyRequest.getDeleteKeys();
+    List<String> deleteKeys = new ArrayList<>(deleteKeyArgs.getKeysList());
+
+    IOException exception = null;
+    OMClientResponse omClientResponse = null;
+    Result result = null;
+
+    OMMetrics omMetrics = ozoneManager.getMetrics();
+    omMetrics.incNumKeyDeletes();
+    String volumeName = deleteKeyArgs.getVolumeName();
+    String bucketName = deleteKeyArgs.getBucketName();
+    Map<String, String> auditMap = new LinkedHashMap<>();
+    auditMap.put(VOLUME, volumeName);
+    auditMap.put(BUCKET, bucketName);
+    List<OmKeyInfo> omKeyInfoList = new ArrayList<>();
+    List<OmKeyInfo> dirList = new ArrayList<>();
+
+    AuditLogger auditLogger = ozoneManager.getAuditLogger();
+    OzoneManagerProtocolProtos.UserInfo userInfo = getOmRequest().getUserInfo();
+
+    OzoneManagerProtocolProtos.OMResponse.Builder omResponse =
+        OmResponseUtil.getOMResponseBuilder(getOmRequest());
+    OMMetadataManager omMetadataManager = ozoneManager.getMetadataManager();
+
+    boolean acquiredLock = false;
+
+    int indexFailed = 0;
+    int length = deleteKeys.size();
+    OzoneManagerProtocolProtos.DeleteKeyArgs.Builder unDeletedKeys =
+        OzoneManagerProtocolProtos.DeleteKeyArgs.newBuilder()
+            .setVolumeName(volumeName).setBucketName(bucketName);
+
+    boolean deleteStatus = true;
+    try {
+      ResolvedBucket bucket =
+          ozoneManager.resolveBucketLink(Pair.of(volumeName, bucketName), this);
+      bucket.audit(auditMap);
+      volumeName = bucket.realVolume();
+      bucketName = bucket.realBucket();
+
+      acquiredLock = omMetadataManager.getLock()
+          .acquireWriteLock(BUCKET_LOCK, volumeName, bucketName);
+      // Validate bucket and volume exists or not.
+      validateBucketAndVolume(omMetadataManager, volumeName, bucketName);
+      String volumeOwner = getVolumeOwner(omMetadataManager, volumeName);
+
+      for (indexFailed = 0; indexFailed < length; indexFailed++) {
+        String keyName = deleteKeyArgs.getKeys(indexFailed);
+        OzoneFileStatus keyStatus = OMFileRequest
```

Comment:
@sadanand48 
I saw the main diff between OMKeysDeleteRequest and OMKeysDeleteRequestWithFSO is on two logic:

1) get OMKeyInfo object.
2) Iterate over dirList and add it to deletedDirectoryTable.
3) getOmClientResponse()

Can we introduce two new methods in OMKeysDeleteRequest class and override these method with specific logics? Basically to unify these two classes and avoid code duplications?

Reply:
@rakeshadr  Too many variables at play here, This would create methods with too many parameters. Are there any references in the code where this is done since there are lot of FSO request classes and most of them have common sanity checks?

Reply:
@sadanand48 Please refer to S3MultipartUploadAbortRequestWithFSO or S3MultipartUploadCommitPartRequestWithFSO or S3MultipartUploadCompleteRequestWithFSO.

Yes, how about to introduce ParameterBuilder to reduce the num of args ?